# Business Context

**A public health clinic aims to understand why many patients fail to attend their scheduled appointments. They want to identify factors that can help them reduce the no-show rate.**

In [9]:
import pandas as pd

https://www.kaggle.com/datasets/joniarroba/noshowappointments

In [10]:
ma = pd.read_csv('KaggleV2-May-2016.csv')

**1. Initial exploration**

- You can use the DataFrame methods from pandas, such as DataFrame.head(), DataFrame.info(), DataFrame.describe(), etc, to investigate the data.
- Alternatively use the following package:

In [11]:
#from pandas_profiling import ProfileReport
#prof = ProfileReport(your_dataframe)
#prof.to_file(output_file='dataset_name.html')

# Basic dataset information
print("Dataset Information:")
print(ma.info())

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB
None


- *ScheduledDay and AppointmentDay are stored as text, should explore as would expect datetime*
- *Scholarship to SMS_Received are stored as int64, should explore whether binary or quantity*
- *No-show is stored as str, should explore what values are stored*

In [12]:
# Header of the dataset
print("Dataset Head:")
print(ma.head())

Dataset Head:
      PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0 

- *Gender seems split as "F" and "M" - is this consistent?* <br>
- *This confirms also that ScheduledDay and AppointmentDay are datetime/dates stored as text* <br>
- *Scholarship to SMS_Received do appear to be binary, but difficult to say with data visible here* <br>
- *No-show does appear to be "No" and maybe "Yes" - this may need further validation* <br>

In [13]:
# Dataset description
print("Dataset Description:")
print(ma.describe())

Dataset Description:
          PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000 

- What are your initial observations about the data?.
<i>
    - There are 110527 records in total.
    - Min age value is -1  - may need cleaning
    - Max age value is 115 - may need cleaning
    - Handicap has max value 4? mean/std are very low, need to examine.
    - Scholarship to SMS_Received are binary values with min/max at 0/1.
</i>


**2. Connect to the business context**

- Based on the business problem of identifying why patients don't show up, which columns are most relevant for your analysis?
    <div><i>
        Initially, I believe the following columns are most relevant:
        
            1.	ScheduleDay – Important to understand if length between schedule day and appointment day 
            2.	AppointmentDay – To find patterns in appointment, common day of the week?
            3.	Sms_received – Receiving sms may influence attendance
            4.	Age – Different age groups may have different attendance patterns
            5.	Neighbourhood – Distance may affect attendance
            6.	No-show – indicates whether attended or not
    </i></div>

- Which column is the final result you are trying to understand or predict?

<div>
        <i>The no-show is the final result I will try to understand/predict. This is the immediate indication of whether a patient attended or not.</i>
</div>

**3. Identify data quality issues**

Identify Data Quality Issues in your selected columns. Now that you've chosen the columns you believe are most relevant, perform a detailed analysis to identify specific issues like missing values, inconsistent formats, incorrect data types, unusual values etc, within those columns only.

In [14]:
# Check for missing values
print("Missing Values:")
print(ma.isnull().sum())

Missing Values:
PatientId         0
AppointmentID     0
Gender            0
ScheduledDay      0
AppointmentDay    0
Age               0
Neighbourhood     0
Scholarship       0
Hipertension      0
Diabetes          0
Alcoholism        0
Handcap           0
SMS_received      0
No-show           0
dtype: int64


*There are no missing values in each of the columns*

In [15]:
# Check for unique values in each column
print("Unique Values in Each Column:")
for column in ma.columns:
    unique_values = ma[column].nunique()
    print(f"{column}: {unique_values} unique values")

Unique Values in Each Column:


PatientId: 62299 unique values
AppointmentID: 110527 unique values
Gender: 2 unique values
ScheduledDay: 103549 unique values
AppointmentDay: 27 unique values
Age: 104 unique values
Neighbourhood: 81 unique values
Scholarship: 2 unique values
Hipertension: 2 unique values
Diabetes: 2 unique values
Alcoholism: 2 unique values
Handcap: 5 unique values
SMS_received: 2 unique values
No-show: 2 unique values


In [16]:
# Show unique values for 'AppointmentDay'
print("Show Appointment Day Unique Actual Values:")
print(ma['AppointmentDay'].unique())

Show Appointment Day Unique Actual Values:
['2016-04-29T00:00:00Z' '2016-05-03T00:00:00Z' '2016-05-10T00:00:00Z'
 '2016-05-17T00:00:00Z' '2016-05-24T00:00:00Z' '2016-05-31T00:00:00Z'
 '2016-05-02T00:00:00Z' '2016-05-30T00:00:00Z' '2016-05-16T00:00:00Z'
 '2016-05-04T00:00:00Z' '2016-05-19T00:00:00Z' '2016-05-12T00:00:00Z'
 '2016-05-06T00:00:00Z' '2016-05-20T00:00:00Z' '2016-05-05T00:00:00Z'
 '2016-05-13T00:00:00Z' '2016-05-09T00:00:00Z' '2016-05-25T00:00:00Z'
 '2016-05-11T00:00:00Z' '2016-05-18T00:00:00Z' '2016-05-14T00:00:00Z'
 '2016-06-02T00:00:00Z' '2016-06-03T00:00:00Z' '2016-06-06T00:00:00Z'
 '2016-06-07T00:00:00Z' '2016-06-01T00:00:00Z' '2016-06-08T00:00:00Z']


<div><i>
It looks like Appointment day covers approx. a 1 month period <br>
There are no times in the datetime string
</i></div>

In [17]:
# Show unique values for 'ScheduledDay'
print("Show Scheduled Day Unique Actual Values:")
print(ma['ScheduledDay'].unique())

Show Scheduled Day Unique Actual Values:
['2016-04-29T18:38:08Z' '2016-04-29T16:08:27Z' '2016-04-29T16:19:04Z' ...
 '2016-04-27T16:03:52Z' '2016-04-27T15:09:23Z' '2016-04-27T13:30:56Z']


*It looks like Scheduled Day includes Date AND Time, whereas Appointment Day does not include time.*

*I will need to exlucde the time element from the data as this will be irrelevant/meaningless.*

In [18]:
# Show unique values for 'Handcap'
print("Show Actual unique values for handicap:")
print(ma['Handcap'].unique())

print("----------------------------------------------------------------------")
print("Handcap value counts:")
print(ma['Handcap'].value_counts())

Show Actual unique values for handicap:
[0 1 2 3 4]
----------------------------------------------------------------------
Handcap value counts:
Handcap
0    108286
1      2042
2       183
3        13
4         3
Name: count, dtype: int64


*It appears the values are possibly a count of the number of handicaps, therefore data seems reasonable*

In [19]:
# Check for duplicated rows
print("Number of duplicated rows:")
print(ma.duplicated().sum())

Number of duplicated rows:


0


*Looks Good*

In [20]:
# Show unique gender values
print("Show Gender unique values:")
print(ma['Gender'].value_counts())

Show Gender unique values:
Gender
F    71840
M    38687
Name: count, dtype: int64


*Somewhat expected - indicates more F than M appontments*

In [21]:
# Show No-show value counts
print("Show No-show value counts:")
print(ma['No-show'].value_counts())

Show No-show value counts:
No-show
No     88208
Yes    22319
Name: count, dtype: int64


In [22]:
# Check for negative ages
print("Check negative ages:")
print(ma[ma['Age'] < 0])

Check negative ages:
          PatientId  AppointmentID Gender          ScheduledDay  \
99832  4.659432e+14        5775010      F  2016-06-06T08:58:13Z   

             AppointmentDay  Age Neighbourhood  Scholarship  Hipertension  \
99832  2016-06-06T00:00:00Z   -1         ROMÃO            0             0   

       Diabetes  Alcoholism  Handcap  SMS_received No-show  
99832         0           0        0             0      No  


*Negative ages should be removed*

In [23]:
# Check for older ages
print("Check Older Ages:")
print(ma[ma['Age'] > 100])  

Check Older Ages:
          PatientId  AppointmentID Gender          ScheduledDay  \
58014  9.762948e+14        5651757      F  2016-05-03T09:14:53Z   
63912  3.196321e+13        5700278      F  2016-05-16T09:17:44Z   
63915  3.196321e+13        5700279      F  2016-05-16T09:17:44Z   
68127  3.196321e+13        5562812      F  2016-04-08T14:29:17Z   
76284  3.196321e+13        5744037      F  2016-05-30T09:44:51Z   
90372  2.342836e+11        5751563      F  2016-05-31T10:19:49Z   
97666  7.482346e+14        5717451      F  2016-05-19T07:57:56Z   

             AppointmentDay  Age Neighbourhood  Scholarship  Hipertension  \
58014  2016-05-03T00:00:00Z  102     CONQUISTA            0             0   
63912  2016-05-19T00:00:00Z  115    ANDORINHAS            0             0   
63915  2016-05-19T00:00:00Z  115    ANDORINHAS            0             0   
68127  2016-05-16T00:00:00Z  115    ANDORINHAS            0             0   
76284  2016-05-30T00:00:00Z  115    ANDORINHAS            0 

*The age 115 may need clearning from the data - as appears to be an anomoly*

In [24]:
# Check for appointments scheduled AFTER the appointment day
print("Check appointments scheduled AFTER the appointment day:")
ma['ScheduledDay_dt'] = pd.to_datetime(ma['ScheduledDay'])
ma['AppointmentDay_dt'] = pd.to_datetime(ma['AppointmentDay'])

invalid = ma[ma['ScheduledDay_dt'] > ma['AppointmentDay_dt']]
print(f"Scheduled after appointment: {len(invalid)}")


Check appointments scheduled AFTER the appointment day:
Scheduled after appointment: 38568


*The time stamps are interfering with the output, I should review by date only*

In [25]:
# Show the invalid rows where ScheduledDay is after AppointmentDay
ma['ScheduledDate'] = ma['ScheduledDay_dt'].dt.date
ma['AppointmentDate'] = ma['AppointmentDay_dt'].dt.date

invalid = ma[ma['ScheduledDate'] > ma['AppointmentDate']]
print(f"Scheduled after appointment: {len(invalid)}")
print(invalid[['ScheduledDay', 'AppointmentDay', 'ScheduledDate', 'AppointmentDate']])

Scheduled after appointment: 5
               ScheduledDay        AppointmentDay ScheduledDate  \
27033  2016-05-10T10:51:53Z  2016-05-09T00:00:00Z    2016-05-10   
55226  2016-05-18T14:50:41Z  2016-05-17T00:00:00Z    2016-05-18   
64175  2016-05-05T13:43:58Z  2016-05-04T00:00:00Z    2016-05-05   
71533  2016-05-11T13:49:20Z  2016-05-05T00:00:00Z    2016-05-11   
72362  2016-05-04T06:50:57Z  2016-05-03T00:00:00Z    2016-05-04   

      AppointmentDate  
27033      2016-05-09  
55226      2016-05-17  
64175      2016-05-04  
71533      2016-05-05  
72362      2016-05-03  


*These will need to be cleaned*

**4. Initial data cleaning and transformation**

(Each person from the group)Choose a relevant column with one or multiple data issues and try to address them. For example: missing values, incorrect data type, mixed types of values, etc.

In [26]:
# I need to remove negative age entries, >110 age entries. 
# I need to convert the "ScheduledDay" and "AppointmentDay" columns to datetime format
# Then remove any appointments that were scheduled after the appointment day.
# Remove negative ages
ma = ma[ma['Age'] >= 0]

# Remove ages above 110
ma = ma[ma['Age'] <= 110]

# Convert columns to datetime format
ma['ScheduledDate'] = pd.to_datetime(ma['ScheduledDate'])
ma['AppointmentDate'] = pd.to_datetime(ma['AppointmentDate'])

# Remove appointments scheduled after the appointment day
ma = ma[ma['ScheduledDate'] <= ma['AppointmentDate']]

print(f"Dataset shape after cleaning: {ma.shape}")
print(f"Remaining records: {len(ma)}")


Dataset shape after cleaning: (110516, 18)
Remaining records: 110516


**5. Bonus task: Early feature engineering**

Discuss and implement the potential feature engineering opportunities in this dataset. What new columns can you create from the existing ones that might enhance the next steps of the AI project?

In [27]:
# Number of days between scheduling and appointment
ma['DaysBetween'] = (ma['AppointmentDate'] - ma['ScheduledDate']).apply(lambda x: x.days)
print("Days between scheduling and appointment:")
print(ma['DaysBetween'].describe())


Days between scheduling and appointment:
count    110516.000000
mean         10.184272
std          15.255201
min           0.000000
25%           0.000000
50%           4.000000
75%          15.000000
max         179.000000
Name: DaysBetween, dtype: float64
